src/Features/Core/Portable/Remote/RemoteArguments.cs

```diff
@@ -172,15 +174,18 @@ public static SerializableNavigableItem Dehydrate(INavigableItem item)
                    item.SourceSpan,
                    item.ChildItems.SelectAsArray(Dehydrate));
 
-        public INavigableItem Rehydrate(Solution solution)
+        public async ValueTask<INavigableItem> RehydrateAsync(Solution solution, CancellationToken cancellationToken)
         {
             var childItems = ChildItems == null
                 ? ImmutableArray<INavigableItem>.Empty
-                : ChildItems.SelectAsArray(c => c.Rehydrate(solution));
+                : ImmutableArray.Create(
+                    await Task.WhenAll(
+                        ChildItems.SelectAsArray(c => c.RehydrateAsync(solution, cancellationToken).AsTask())).ConfigureAwait(false));
```

Comment:
💭 I'm not a huge fan of this over iterating them and avoiding the task allocations, _especially_ if they are generally expected to complete synchronously.

Reply:
@sharwell Suggestions? If it doesn't already exist I could write a helper that essentially iterates the ValueTasks directly, and only falls to the Task.WhenAll() path once it finds the first one that isn't already completed synchronously. I agree I don't like this, I wasn't sure I wanted to write a bunch of icky complexity though if avoidable.

Reply:
I would suggest creating a `SelectAsArrayAsync` extension, which returns a `ValueTask<ImmutableArray<T>>` and operates using a `ValueTask`-returning transformation function.

Reply:
@sharwell: https://github.com/dotnet/roslyn/pull/50662/commits/9dea92725ec9fbbf10d3acac521f72767507e2fb